### PR TITLE
#28 user-register

### LINKED ISSUE #28: Feat: Register User (Vendors & Buyers)
## Description

As a vendor or buyer, I want to be able to register an account on the e-commerce platform so that I can access the respective features and functionalities.

The registration process should allow users to sign up as either a vendor or a buyer, and capture relevant information such as name, email, password, and any additional details required for each user type.

### Acceptance Criteria

- [ ] Implement a user registration API endpoint
- [ ] Validate user input (e.g., email, password, user type)
- [ ] Check for existing users (email uniqueness)
- [ ] Hash and store user passwords securely
- [ ] Create user records in the database with appropriate user types (vendor or buyer)
- [ ] Return appropriate success or error responses
- [ ] Implement input sanitization

## Developer Notes

- Design the user registration data model and database schema
- Define the API endpoint and request/response payloads
- Implement input validation and sanitization measures
- Use a secure password hashing algorithm (e.g., bcrypt, argon2)
- Implement email user verification
- Write unit tests to cover different scenarios and edge cases
- Write clear API documentation for resulting endpoints

## Related Issues

- #xx (User Login)
- #xx (User Profile Management)
- #xx (Database Schema Design)

## Definition of Done

- User registration API endpoint is implemented and functional
- User input validation and sanitization measures are in place
- Passwords are securely hashed and stored
- User records are created in the database with appropriate user types
- Appropriate success and error responses are returned
- Unit tests are written and passing
- Code adheres to coding standards and best practices
- All created endpoints are documented

### PR DESCRIPTION
**Pull request by ambroisegithub**
**Pull Request Description:**

This pull request addresses the implementation of user registration functionality for the e-commerce platform. It allows vendors or buyers to register an account, enabling access to platform features.

**Description of Task to be completed:**

As a user, I should be able to register an account on the e-commerce platform with ease. The tasks included in this **pull request are as follows:**

-  [ ]  Implement a user registration API endpoint: Develop an endpoint that facilitates user registration, capturing necessary information like name, email, password, and user type (vendor or buyer).
- [ ]  Validate user input: Ensure validation of user input fields such as email format, password strength, and user type.
- [ ] Check for existing users: Validate the uniqueness of email addresses to prevent duplicate registrations.
- [ ] Hash and store user passwords securely: Utilize a secure password hashing algorithm (e.g., bcrypt) to securely store user passwords in the database.
- [ ] Create user records in the database: Upon successful registration, create user records in the database with appropriate user types.
- [ ] Return appropriate responses: Implement response handling to return appropriate success or error responses based on registration outcomes.
- [ ] Implement input sanitization: Apply input sanitization measures to prevent injection attacks and ensure data integrity.

**manually test the application after:**

-  **Build the Project**

        npm run build

-  **start Project**

        npm run start

-   **To test Project**

        npm run test

**Access API Documentation (Swagger):**

Open a web browser and navigate to the appropriate URL where the Swagger API documentation is hosted. This URL should be provided in the project documentation or README. For example:

**http://localhost:3000/api-docs**

**Developer Notes:**

**To accomplish the above tasks effectively, the following steps need to be taken:**

-  Design the user registration data model and database schema.
-  Define the API endpoint structure along with request/response payloads.
-  Implement input validation and sanitization mechanisms to enhance security.
-  Utilize a secure password hashing algorithm like bcrypt to safeguard user passwords.
-  Incorporate email user verification for added security measures.
-  Write comprehensive unit tests covering various scenarios and edge cases.
-  Document resulting endpoints clearly in the API documentation for user reference.

**Related Issues:**

**This pull request addresses the following related issues:**

#29 
#35 

**Definition of Done:**

**The completion criteria for this task include:**

- Implementation and functionality of the user registration API endpoint.
- Implementation of user input validation and sanitization measures.
- Secure hashing and storage of user passwords.
- Creation of user records in the database with appropriate user types.
- Proper handling of success and error responses.
- Writing and passing of unit tests.
- Adherence to coding standards and best practices.
- Clear documentation of all created endpoints.
- 
**Labels:**

feature

### CODE REVIEW

**Review by wayneleon1 — APPROVED** [review]
That looks good to me

**Review by bertrandshema — CHANGES_REQUESTED** [review]
In the Models folder we have 2 files, one called user.ts, and the other one called userModel.ts. those 2 files are doing the same thing which is defining the user model, please choose one to continue with. 

In terms of naming since the files are in the models folder, TTL mentioned to me that do not need to add model on each file for example(userModel, productModel) you can just simply call them user or products as the folder they are in specifies whether if they are models or controller, etc.. but still it is your choice on how you will reneme

**Review by EddyShimwa — APPROVED** [review]

**Review by EddyShimwa — APPROVED** [review]

**Comment by ambroisegithub** [comment]
> In the Models folder we have 2 files, one called user.ts, and the other one called userModel.ts. those 2 files are doing the same thing which is defining the user model, please choose one to continue with.
> 
> In terms of naming since the files are in the models folder, TTL mentioned to me that do not need to add model on each file for example(userModel, productModel) you can just simply call them user or products as the folder they are in specifies whether if they are models or controller, etc.. but still it is your choice on how you will reneme

**Review by bertrandshema — APPROVED** [review]
It is all good for me now

File: .eslintrc.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
   },
   "rules": {
     "no-console": "warn",
-    "quotes": ["error", "single"]
+    "@typescript-eslint/no-var-requires": "off",
```
Comment (Habinezajanvier) [inline .eslintrc.json]: Please, don't put `no-var-requires` off

Comment (ambroisegithub) [reply .eslintrc.json]: When I remove this can cause the ci/cd eslintsr errors, let me try to remove them and see what will be happen.

Comment (ambroisegithub) [reply .eslintrc.json]: Ok Done This was made It:

 "ignorePatterns": ["dist/**/*"]

**Review by Habinezajanvier — APPROVED** [review]
:+1:
